traffic_router/connector/src/main/java/com/comcast/cdn/traffic_control/traffic_router/secure/CertificateRegistry.java

```diff
@@ -37,9 +42,31 @@ private CertificateRegistry() {
 	}
 
 	public static CertificateRegistry getInstance() {
+		synchronized (CertificateRegistryHolder.DELIVERY_SERVICE_CERTIFICATES) {
```

Comment:
Yes. The `importCertificateData` also synchronizes on the `CertificateRegistry` instance.  Also, the `getInstance()` method could be called by multiple threads.

Reply:
`importCertificateDataList` is a synchronized method, which I think is the same as if the method body was synchronized on `CertificateRegistry.class`, right?

Reply:
Yes. I placed this Synchronize here to protect the data structure in the event that more than one thread calls the 'getInstance' method during startup. There are in fact three endpoints being initialized using the LanguidNioProtocol at startup. Once initialized this method never gets called again so they synchronized doesn't have any impact on performance.

Reply:
The data structure would only be protected from race conditions if the other code that accesses it also synchronized on the same object. `synchronized (CertificateRegistryHolder.DELIVERY_SERVICE_CERTIFICATES)` and `synchronized (CertificateRegistry.class)` are locking on two different things, so both of those methods could overlap with each other.

Reply:
Basically, I'm trying to understand why this method isn't just `synchronized public void` like `importCertificateDataList` if it's also meant to not overlap with `importCertificateDataList`. That method is also mutating `handshakeDataMap`, so I'd think `getInstance` and `importCertificateDataList` would have to be sync'd on the same object (i.e. use the same mutex).

Reply:
I've tested with your suggestion and it works. I was concerned that locking the entire class, (since getInstance is a static method) might prevent the CertificateRegistryHolder from creating the new CertificateRegistry, but it looks like the class loader is getting to the CertificateRegistryHolder first.

Reply:
this is done in new commits

Reply:
Ah yeah I just realized that `getInstance` is now `static synchronized` while `importCertificateDataList` is non-static `synchronized`. So those methods aren't actually locking on the same object yet (`static` locks on the class while non-static locks on a single instance).

So if one thread calls `getInstance` and another thread calls `importCertificateDataList` at the same time, both methods will interleave and mutate `handshakeDataMap` at the same time. That's what you're trying to prevent with the synchronization here, right?

Reply:
I'm more concerned about simultaneous calls to 'getInstance' during startup. The poller that calls 'importCertificateDataList' launches after initialization has succeeded.

Reply:
This discussion has caused me to realize that there is a bug in `getInstance`. I've just committed another change.